plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,208 @@
+
+mkSignatureHelp :: Name -> Text -> UInt -> SignatureHelp
+mkSignatureHelp functionName functionType argumentNumber =
+    let functionNameLabelPrefix = printOutputable (ppr functionName) <> " :: "
+     in SignatureHelp
+            [ SignatureInformation
+                  (functionNameLabelPrefix <> functionType)
+                  Nothing
+                  (Just $ mkArguments (fromIntegral $ T.length functionNameLabelPrefix) functionType)
+                  (Just $ InL argumentNumber)
+            ]
+            (Just 0)
+            (Just $ InL argumentNumber)
+
+-- TODO(@linj) can type string be a multi-line string?
+mkArguments :: UInt -> Text -> [ParameterInformation]
+mkArguments offset functionType =
+    let separator = " -> "
+        separatorLength = fromIntegral $ T.length separator
+        splits = T.breakOnAll separator functionType
+        prefixes = fst <$> splits
+        prefixLengths = fmap (T.length >>> fromIntegral) prefixes
+        ranges =
+            [ ( if previousPrefixLength == 0 then 0 else previousPrefixLength + separatorLength,
+                currentPrefixLength
+              )
+            | (previousPrefixLength, currentPrefixLength) <- zip (0: prefixLengths) prefixLengths
+            ]
+     in [ ParameterInformation (InR range) Nothing
+        | range <- bimap (+offset) (+offset) <$> ranges
+        ]
+
+extractInfoFromSmallestContainingFunctionApplicationAst ::
+    Position -> HieASTs a -> (RealSrcSpan -> HieAST a -> Maybe b) -> [b]
+extractInfoFromSmallestContainingFunctionApplicationAst position hieAsts extractInfo =
+    M.elems $ flip M.mapMaybeWithKey (getAsts hieAsts) $ \hiePath hieAst ->
+        smallestContainingSatisfying (positionToSpan hiePath position) (nodeHasAnnotation ("HsApp", "HsExpr")) hieAst
+            >>= extractInfo (positionToSpan hiePath position)
+    where
+        positionToSpan hiePath position =
+            let loc = mkLoc hiePath position in mkRealSrcSpan loc loc
+        mkLoc (LexicalFastString hiePath) (Position line character) =
+            mkRealSrcLoc hiePath (fromIntegral line + 1) (fromIntegral character + 1)
+
+type Annotation = (FastStringCompat, FastStringCompat)
+
+nodeHasAnnotation :: Annotation -> HieAST a -> Bool
+nodeHasAnnotation annotation = sourceNodeInfo >>> maybe False (isAnnotationInNodeInfo annotation)
+
+-- TODO(@linj): the left most node may not be the function node. example: (if True then f else g) x
+getLeftMostNode :: HieAST a -> HieAST a
+getLeftMostNode thisNode =
+    case nodeChildren thisNode of
+        []           -> thisNode
+        leftChild: _ -> getLeftMostNode leftChild
+
+getNodeName :: RealSrcSpan -> HieAST a -> Maybe Name
+getNodeName _span hieAst =
+    if nodeHasAnnotation ("HsVar", "HsExpr") hieAst
+        then
+            case mapMaybe extractName $ M.keys $ M.filter isUse $ getSourceNodeIds hieAst of
+                [name] -> Just name -- TODO(@linj) will there be more than one name?
```

Comment:
I'm not 100% sure when/if this can happen. I think it might happen if we're looking at a typeclass method, in which case we could get the generic typeclass method and the specific one perhaps? That would be a good use-case for multilple signatures, e.g. if we have 

```
foldMap @[] f ^
```
then we might want to see both the generic `Foldable` signature and the one for the list instance.

Reply:
I think my general suggestion would be: let's treat possibly-multiple node annotations as producing possibly-multiple signatures.

Reply:
Implemented in d603ec4a3008983bf5f70e6ff63a30d9d8387c06.

Related tests:
- type constraint
- type constraint with kind signatures
- multi-line type with type constraint

---

> I'm not 100% sure when/if this can happen.

I do not have a principled conclusion.  But from my experiments, the summary is that
- There are multiple types iif there are type variables.
- There seems to be only one `Name`.